### PR TITLE
ASIM Audit schema parser with its sample and test data for Barracuda …

### PR DESCRIPTION
**Pull request by jayeshprajapaticrest**
Required items, please complete
   
   Change(s):
   - Added ASIM Audit schema parser for Barracuda WAF

   Reason for Change(s):
   - Initial version ASIM Audit Parser

   Version Updated:
   - Initial version

   Testing Completed:
   - Yes

   Checked that the validations are passing and have addressed any issues that are present:
   - Yes

### CODE REVIEW

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventBarracudaWAF.yaml [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+      | extend 
+          severity = toint(Severity_s)
+      | lookup SeverityLookup on severity
+      | lookup EventTypeLookup on ChangeType_s
+      | extend
+          EventResult = "Success", 
+          EventSchema = "AuditEvent",
+          EventSchemaVersion = "0.1.0",
+          EventVendor = "Barracuda",
+          EventProduct = "WAF"
```
Comment (devikamehra) [inline Parsers/ASimAuditEvent/Parsers/ASimAuditEventBarracudaWAF.yaml:59]: EventCount is missing

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventBarracudaWAF.yaml [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+      | extend
+          Dvc = column_ifexists("UnitName_s", ""), 
+          EventStartTime = unixtime_milliseconds_todatetime(tolong(DeviceReceiptTime_s)),
+          Operation = column_ifexists("CommandName_s", ""),
```
Comment (devikamehra) [inline Parsers/ASimAuditEvent/Parsers/ASimAuditEventBarracudaWAF.yaml:63]: Variable_s, ObjectName_s and ObjectType_s is missing the mapping.
Object should be Variable_s/ObjectName_s
ObjectType should be based on ObjectType_s according to the allowed values.

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventBarracudaWAF.yaml [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+          DvcIpAddr = column_ifexists("HostIP_s", ""),
+          NewValue = column_ifexists("NewValue_s", ""),
+          Src = column_ifexists("LoginIP_s", ""),
```
Comment (devikamehra) [inline Parsers/ASimAuditEvent/Parsers/ASimAuditEventBarracudaWAF.yaml:66]: SrcIpAddr should be assigned the value. Src should only be an alias.

File: Parsers/ASimAuditEvent/Parsers/vimAuditEventBarracudaWAF.yaml [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+      newvalue_has_any: dynamic=dynamic([]),
+      operation_has_any: dynamic=dynamic([])) {
+      barracuda_CL
+      | where not(disabled) 
+          and LogType_s == "AUDIT" 
+          and EventName_s !in ("LOGIN", "LOGOUT", "UNSUCCESSFUL_LOGIN")
+      | where (array_length(srcipaddr_has_any_prefix) == 0 or has_any_ipv4_prefix(LoginIP_s, srcipaddr_has_any_prefix))
+      | extend EventStartTime = unixtime_milliseconds_todatetime(tolong(DeviceReceiptTime_s))
+      | extend EventEndTime = EventStartTime
+      | where (isnull(starttime) or EventStartTime >= starttime) and (isnull(endtime) or EventEndTime <= endtime)
```
Comment (devikamehra) [inline Parsers/ASimAuditEvent/Parsers/vimAuditEventBarracudaWAF.yaml:82]: Move this before ip based filter.
Ensure all the variables have column_ifexists.

File: Parsers/ASimAuditEvent/Tests/Barracuda_WAF_ASimAuditEvent_DataTest.csv [hunk 5]
```diff
@@ -0,0 +1,3 @@
+﻿Result
+"(0) Error: 1 invalid value(s) (up to 10 listed) in 1000 records (100.0%) for field [EventProduct] of type [Enumerated]: [""WAF""] (Schema:AuditEvent)"
+"(0) Error: 1 invalid value(s) (up to 10 listed) in 1000 records (100.0%) for field [EventVendor] of type [Enumerated]: [""Barracuda""] (Schema:AuditEvent)"
```
Comment (devikamehra) [inline Parsers/ASimAuditEvent/Tests/Barracuda_WAF_ASimAuditEvent_DataTest.csv:3]: Are these results complete? No warnings were received?

File: Parsers/ASimAuditEvent/Tests/Barracuda_WAF_ASimAuditEvent_SchemaTest.csv [hunk 6]
```diff
@@ -0,0 +1,98 @@
+﻿Result
+"(1) Warning: Missing recommended field [ActorUsername]"
```
Comment (devikamehra) [inline Parsers/ASimAuditEvent/Tests/Barracuda_WAF_ASimAuditEvent_SchemaTest.csv:2]: AdminName_s is present and not mapped to ActorUsername. Please map ActorUsernameType and ActorUserType as well.

File: Parsers/ASimAuditEvent/Tests/Barracuda_WAF_ASimAuditEvent_SchemaTest.csv [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+"(1) Warning: Missing recommended field [Dst]"
+"(1) Warning: Missing recommended field [DvcAction]"
+"(1) Warning: Missing recommended field [DvcDomain]"
+"(1) Warning: Missing recommended field [EventResultDetails]"
```
Comment (devikamehra) [inline Parsers/ASimAuditEvent/Tests/Barracuda_WAF_ASimAuditEvent_SchemaTest.csv:6]: Reason in EventMessage should be mapped to EventResultDetails.

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventBarracudaWAF.yaml [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+          EventMessage = column_ifexists("EventMessage_s", ""),
+          OldValue = column_ifexists("OldValue_s", ""),
+          DvcHostname = column_ifexists("host_s", "")
+      | extend
+          SrcIpAddr = Src,
+          EventEndTime=EventStartTime
+      | extend
+          IpAddr = SrcIpAddr,
+          Value = NewValue
```
Comment (devikamehra) [inline Parsers/ASimAuditEvent/Parsers/ASimAuditEventBarracudaWAF.yaml:75]: ValueType is missing

**Review by devikamehra — DISMISSED** [review]

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventBarracudaWAF.yaml (hunk 2 above)
Comment (jayeshprajapaticrest) [reply Parsers/ASimAuditEvent/Parsers/ASimAuditEventBarracudaWAF.yaml]: Done

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventBarracudaWAF.yaml (hunk 3 above)
Comment (jayeshprajapaticrest) [reply Parsers/ASimAuditEvent/Parsers/ASimAuditEventBarracudaWAF.yaml]: Done

File: Parsers/ASimAuditEvent/Parsers/vimAuditEventBarracudaWAF.yaml (hunk 4 above)
Comment (jayeshprajapaticrest) [reply Parsers/ASimAuditEvent/Parsers/vimAuditEventBarracudaWAF.yaml:125]: Done
We removed column_ifexists and used the the approach shared by you in mail to create one data table with the default values.

File: Parsers/ASimAuditEvent/Tests/Barracuda_WAF_ASimAuditEvent_DataTest.csv (hunk 5 above)
Comment (jayeshprajapaticrest) [reply Parsers/ASimAuditEvent/Tests/Barracuda_WAF_ASimAuditEvent_DataTest.csv]: Yes

File: Parsers/ASimAuditEvent/Tests/Barracuda_WAF_ASimAuditEvent_SchemaTest.csv (hunk 6 above)
Comment (jayeshprajapaticrest) [reply Parsers/ASimAuditEvent/Tests/Barracuda_WAF_ASimAuditEvent_SchemaTest.csv]: Done

File: Parsers/ASimAuditEvent/Tests/Barracuda_WAF_ASimAuditEvent_SchemaTest.csv (hunk 7 above)
Comment (jayeshprajapaticrest) [reply Parsers/ASimAuditEvent/Tests/Barracuda_WAF_ASimAuditEvent_SchemaTest.csv]: Done

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventBarracudaWAF.yaml (hunk 8 above)
Comment (jayeshprajapaticrest) [reply Parsers/ASimAuditEvent/Parsers/ASimAuditEventBarracudaWAF.yaml]: Done

**Comment by v-atulyadav** [comment]
Hi @jayeshprajapaticrest, please resolve branch conflicts. Thanks

**Comment by v-atulyadav** [comment]
Hi @jayeshprajapaticrest, please update your branch by taking the latest from master to resolve the conflicts. Thanks

**Comment by v-atulyadav** [comment]
Hi @jayeshprajapaticrest, 
kql validations failed please see below screenshot for the same. Thanks
image

**Comment by v-atulyadav** [comment]
Hi @jayeshprajapaticrest, please act on validation failures. Thanks

**Comment by jayeshprajapaticrest** [comment]
@v-atulyadav Fixed.

**Comment by v-atulyadav** [comment]
Hi @jayeshprajapaticrest, please update your branch by taking the latest from master to resolve the conflicts. Thanks

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventBarracudaWAF.yaml [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+          _ResourceId,
+          SentBytes,
+          ReportReferenceLink,
+          Computer,
+          TenantId
+  }
+  ;
+  union isfuzzy = true 
+      BarracudaCustom(disabled=disabled),
+      BarracudaCEF(disabled=disabled)
```
Comment (vakohl) [inline Parsers/ASimAuditEvent/Parsers/ASimAuditEventBarracudaWAF.yaml:236]: @jayeshprajapaticrest  Why we added filtering to these functions separately?  We should do union under a parent function and add filter to that parent function. Please refer this : https://github.com/Azure/Azure-Sentinel/blob/master/Parsers/ASimNetworkSession/Parsers/ASimNetworkSessionCiscoMeraki.yaml

File: Parsers/ASimAuditEvent/Parsers/vimAuditEventBarracudaWAF.yaml [hunk 10]
```diff
@@ -0,0 +1,307 @@
...
+      BarracudaCEF(
+          disabled=disabled,
+          starttime=starttime,
+          endtime=endtime,
+          srcipaddr_has_any_prefix=srcipaddr_has_any_prefix,
+          eventtype_in=eventtype_in,
+          eventresult=eventresult,
+          newvalue_has_any=newvalue_has_any,
+          operation_has_any=operation_has_any
+      )
```
Comment (vakohl) [inline Parsers/ASimAuditEvent/Parsers/vimAuditEventBarracudaWAF.yaml:307]: @jayeshprajapaticrest Why we added filtering to these functions separately? We should do union under a parent function and add filter to that parent function. Please refer this : https://github.com/Azure/Azure-Sentinel/blob/master/Parsers/ASimNetworkSession/Parsers/ASimNetworkSessionCiscoMeraki.yaml

This is applicable to all Barracuda PRs. Please update

Comment (vakohl) [reply Parsers/ASimAuditEvent/Parsers/vimAuditEventBarracudaWAF.yaml:307]: applicable to Authentication, Network and Web PR

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventBarracudaWAF.yaml (hunk 9 above)
Comment (jayeshprajapaticrest) [reply Parsers/ASimAuditEvent/Parsers/ASimAuditEventBarracudaWAF.yaml]: Done

File: Parsers/ASimAuditEvent/Parsers/vimAuditEventBarracudaWAF.yaml (hunk 10 above)
Comment (jayeshprajapaticrest) [reply Parsers/ASimAuditEvent/Parsers/vimAuditEventBarracudaWAF.yaml]: Done in all Barracuda PRs. Authentication, Network and Web PR

**Comment by v-atulyadav** [comment]
Hi @jayeshprajapaticrest,  
Thank you for accommodating the suggested changes! We will review the Pull Request internally and get back to you by shortly.

**Review by vakohl — APPROVED** [review]

**Review by v-atulyadav — APPROVED** [review]

**Pull request merged**